File: 210/level3/디스크컨트롤러.java
// 20120912 :: SFJ

package level3;

import java.util.*;

class Main {
    public static void main(String[] args) throws Exception {
        int jobs[][] = {{0, 3}, {1, 9}, {2, 6}};

        Solution sol = new Solution();
        System.out.println("result : " + sol.solution(jobs));
    }
}


class Solution {
    public int solution(int[][] jobs) {
        int length = jobs.length;
		int answer = 0;
		int end = 0;
		int currJob = 0;
        int cnt = 0;

        // 2차원 배열 정렬
        Arrays.sort(jobs, (o1, o2) -> o1[0] - o2[0]);   // 2차원 배열 정렬 보기
		PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);

		// 모든 작업에 대해서
		while (cnt < length) {
			while (currJob < length && jobs[currJob][0] <= end) {
				pq.add(jobs[currJob++]);
			}

            if (pq.isEmpty()) {
                end = jobs[currJob][0];
            }

            else {
                int[] tmp = pq.poll();
                end += tmp[1];
				answer += tmp[1] + end - tmp[0];
				cnt++;
			}
		}

		return (int) answer / length;
	}
}

File: JunJangE/level3/디스크 컨트롤러.kt
package level3

import java.util.PriorityQueue

class Solution {
    fun solution(jobs: Array<IntArray>): Int {
        var answer = 0
        val pq : PriorityQueue<IntArray> = PriorityQueue<IntArray>(compareBy { it[0] }) // 작업의 소요시간 기준으로 정렬
        var now : Int = 0
        var start : Int = -1
        var n : Int = 0

        while (n < jobs.size){
            // 반복문을 통해 현재 작업이 요청되었고 현재 시간 이전에 끝날 수 있는 작업을 큐에 담는다.
            for ((i, j) in jobs){
                if (i in (start + 1)..now){
                    pq.add(intArrayOf(j, i))
                }
            }

            if (pq.size > 0){
                val cur = pq.poll()
                start = now
                now += cur[0]
                answer += now - cur[1] // 작업 요청시간부터 종료시간까지 계산
                n++
            } else{
                now++
            }


        }

        return answer / jobs.size
    }
}

fun main(){

  println(Solution1().solution(arrayOf(intArrayOf(0,3), intArrayOf(1,9), intArrayOf(2,6))))
}
